packages/@lwc/engine/src/framework/api.ts

```diff
@@ -347,10 +359,24 @@ export function c(
     Ctor: ComponentConstructor,
     data: CustomElementCompilerData,
     children?: VNodes
-): VCustomElement {
+): VCustomElement | VElement {
     if (isCircularModuleDependency(Ctor)) {
         Ctor = resolveCircularModuleDependency(Ctor);
     }
+    if (isFunction(Ctor) && Ctor.prototype instanceof HTMLElement) {
```

Comment:
this is the expensive part compared to what we had before, these two checks will be ran for every vnode created for a custom element, that's not a lot, but still.

Reply:
No way in hell we will be doing an instanceOf check in the critical path.

Reply:
an alternative here is to use:

```
if (isFunction(Ctor) && !isUndefined(Ctor.prototype.DOCUMENT_NODE) && Ctor.prototype instanceof HTMLElement) {
```

Since `LightningElement` does not extend `HTMLElement `, this numeric check is going to cover 99% of the cases, and the instanceof check will cover the web components cases, which I'm fine with the perf penalty on that case. 

Btw, I don't see a perf penalty at the micro-level so far.

Reply:
First of all you are making a lot of assumptions, like the resolver will point and resolve from a magical npm. Second the whole point of this proposal should be to allow the semantics of WC which include a upgrade at any point it time, which mean we should not enforce a web component to be present at evaluation time.

Reply:
I'm explicitly going against that possibility @diervo, I don't think we should allow that in LWC. The premise in LWC is that your deps are always present by the time you render your component, and we should preserve that, even when it comes to WC.